add server side sampling docs (#5315)
getsentry/sentry-docs: src/docs/product/data-management-settings/server-side-sampling/index.mdx

---
title: Server-side Sampling
sidebar_order: 1
description: "Learn about how you can control which transactions Sentry retains."
---

> Server-side sampling lets you control what transactions Sentry retains by setting sampling rules and rates so you see more of the transactions you want to explore further, and less of the ones you don’t, without reconfiguring the SDK or redeploying your application.

Reviewer: > further, and less of the ones you don’t, without reconfiguring the SDK or redeploying your application.

I think we should make the connection to client-side and explicitly explain the difference to client-side sampling in the end.

"further, and less of the ones you don’t. In contrast to client-side sampling [LINK], server-side sampling lets you manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and to redeploy your application."

Feel free to tweak.

Reviewer: This first paragraph still doesn't answer the "why" completely. 
As a user, I might end up here and may not know why I should care about this in the first place.
Sampling mainly exists to balance data usage and data fidelity so that you get everything that matters without storing too many likely redundant events.

I think that what we have in the [JavaScript docs](https://docs.sentry.io/platforms/javascript/configuration/sampling/#sampling-transaction-events) captures this pretty well.

> Server-side sampling lets you control what transactions Sentry retains by setting sampling rules and rates so you see more of the transactions you want to explore further, and less of the ones you don’t, without reconfiguring the SDK or redeploying your application.

Reviewer: I think we should make the connection to client-side and explicitly explain the difference to client-side sampling in the end. feel free to tweak for grammar and language.

Feel free to tweak.
```suggestion
Server-side sampling lets you control what transactions Sentry retains by setting sampling rules and rates so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to client-side sampling [LINK], server-side sampling lets you manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and to redeploy your application.
```

![The Server-side Sampling page with a list of rules](server-side-sampling.png)

Because every project is different, you can set multiple sample rules with different sample rates per project. For example, you may need more events from high converting pages, critical API endpoints, or need to focus on latency issues from your latest release.

> Server-side sampling also makes your SDK and server-side sampling largely independent of each other. This helps if you're building native applications with many releases active at a given time, or you've already configured server-side sampling to target specific transactions, but the SDK blocks more of those transactions being sent to our servers.

Reviewer: ```suggestion
Server-side sampling also makes your SDK and your sampling configuration largely independent of each other. This helps if you're building native applications with many releases active at a given time. 
```

Reviewer 2: I agree, I think this simplification makes sense and we also avoid repeating server-side too many times in one paragraph.

> Server-side sampling also makes your SDK and server-side sampling largely independent of each other. This helps if you're building native applications with many releases active at a given time, or you've already configured server-side sampling to target specific transactions, but the SDK blocks more of those transactions being sent to our servers.

Reviewer: ```suggestion
Server-side sampling also makes your SDK and server-side sampling largely independent of each other. This helps if you're building native applications with many releases active at a given time, or you've already configured server-side sampling to target specific transactions, but the SDK blocks more of those transactions being sent to our servers.
```
This part is a bit hard to understand.

## How Sentry Samples

> Let's imagine for a moment that the total number of transactions the SDK observes is a bucket of rocks. The way Sentry samples transactions is similar to a person sifting rocks: by pouring the bucket of rocks into a sieve, you're allowing rocks small enough to pass through it:

Reviewer: I think when we talk to the total of things, we have to refer to traces instead of transactions here as the 1000 is traces. 

```suggestion
Let's imagine for a moment that the total number of traces the SDK observes is a bucket of rocks. The way Sentry samples traces is similar to a person sifting rocks: by pouring the bucket of rocks into a sieve, you're allowing rocks small enough to pass through it:
```

Reviewer 2: Yes, I agree that in this case it might be better to refer to traces, rather than transactions.

Author: Hmmm, not sure this isn't going to be more confusing without updating the UI. The feature in the app primarily talks about setting transaction sampling rates, not traces sampling rates. I think we might be introducing complexity in a way that doesn't actually improve comprehension, if that makes sense.

| Sampling Process | Rate | Count |
|------------------|------|-------|
|Total transactions observed | N/A | 1000 |
| Sentry SDK sampled | 25% | 250 |
| Sentry servers sampled | 50% | 250 |

<!-- prettier-ignore-end -->

## Traces & Propagation of Sampling Decisions

TBA

## Server-Side Sampling Configurations

<!-- TO DO: Decide if this should be a child page or remain on this page -->

Now that you know how Sentry is sampling at a high-level, you can set server-side sampling rules in **[Project] > Settings > Server-side Sampling**.

### Order of Precedence

Rules evaluate from top to bottom, with the first matching rule determining the sample rate. Like with declaring a conditional statement in code, rules with the most specific conditional clauses should be evaluated first (`If`), followed by more generic ones (`Else If`). If generic rules lead the list, the more specific rules will likely never be evaluated. Setting a uniform sampling rate is equivalent to an `Else` clause which will always trail rules with conditional clauses.

The Server-side Sampling page in sentry.io allows you to change the order of your rules so you can set it effectively.

> ### Good Examples

Reviewer: I like this in general, but shouldn't we point out that this is just pseudo code as the rules are set via UI?
I may get confused wondering where to put that.

**Most specific rule leads the list**

The rule with the most specific conditional clause will always be evaluated because it's first in order.

```

> if (release = 3.0)  && (environment = production) {

Reviewer 3: ```suggestion
if (release = 3.0) && (environment = production) {
```

    transactionSample = 1.0
} else if (environment = production) {
    transactionSample = 0.5
} else {

>   transactionSample = 0.25

Reviewer 3: ```suggestion
    transactionSample = 0.25
```

}
```

**Each rule is evaluated independently**

This configuration also works just fine because each rule will be evaluated based on entirely different conditions.

```

> if (release = 3.0)  {

Reviewer 3: ```suggestion
if (release = 3.0) {
```

    transactionSample = 1.0
} else if (environment = production) {
    transactionSample = 0.5
} else {

>   transactionSample = 0.25

Reviewer 3: ```suggestion
    transactionSample = 0.25
```

}
```

### Poor Example

The rule with the most specific conditional clause will never be evaluated because a rule with a similar, but more generic clause precedes it.

```
if (environment = production) {
    transactionSample = 1.0

> } else if (release = 3.0)  && (environment = production) {

Reviewer 3: ```suggestion
} else if (release = 3.0) && (environment = production) {
```

    transactionSample = 0.5 // Will never be evaluated
} else {

>   transactionSample = 0.25

Reviewer 3: ```suggestion
    transactionSample = 0.25
```

---
title: Server-side Sampling
sidebar_order: 1
description: "Learn about how you can control which transactions Sentry retains."
---

> 

Author: ```suggestion

<Note>

This feature is in beta and is only available if your organization is participating in its limited release. Features in beta are still in-progress and may have bugs. We recognize the irony. If you’re interested in joining this limited release, please reach out at [abc@sentry.io](mailto:abc@sentry.io).

</Note>

```

> Server-side sampling lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.

Author: ```suggestion
Capturing a single trace involves minimal overhead, but capturing traces for every page load or every API request may add an undesirable load to your system. Sampling your events allows you to better manage the number of events sent to Sentry, so you can tailor your volume to your organization's needs. _Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.
```

---
title: Server-side Sampling
sidebar_order: 1
description: "Learn about how you can control which transactions Sentry retains."
---

<Note>

This feature is in beta and is only available if your organization is participating in its limited release. Features in beta are still in-progress and may have bugs. We recognize the irony. If you’re interested in participating, [join the waitlist](https://sentry.io/for/performance/#updates-signup).

</Note>

> Capturing a single trace involves minimal overhead, but capturing traces for every page load or every API request may add an undesirable load to your system. Sampling your events allows you to better manage the number of events sent to Sentry, so you can tailor your volume to your organization's needs. _Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.

Author: ```suggestion
Capturing a single trace involves minimal overhead, but capturing traces for every page load or every API request may add an undesirable load to your system. Sampling your events allows you to better manage the number of events sent to Sentry, so you can tailor your volume to your organization's needs. 

_Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.
```

> Capturing a single trace involves minimal overhead, but capturing traces for every page load or every API request may add an undesirable load to your system. Sampling your events allows you to better manage the number of events sent to Sentry, so you can tailor your volume to your organization's needs. _Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.

Author: > This first paragraph still doesn't answer the "why" completely.
> As a user, I might end up here and may not know why I should care about this in the first place.
> Sampling mainly exists to balance data usage and data fidelity so that you get everything that matters without storing too many likely redundant events.
> 
> I think that what we have in the [JavaScript docs](https://docs.sentry.io/platforms/javascript/configuration/sampling/#sampling-transaction-events) captures this pretty well.

@danielkhan & @ale-cota - I've tried to update this per the JS docs. Feel free to tweak.

_Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.

![The Server-side Sampling page with a list of rules](server-side-sampling.png)

> 

Author: ```suggestion

> [A trace](/product/sentry-basics/tracing/) A trace represents the record of the entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and it can be distributed across multiple services or projects.

,  

```

|Total transactions observed | N/A | 1000 |
| Sentry SDK sampled | 50% | 500 |
| Sentry servers sampled | 25% | 250 |

<!-- prettier-ignore-end -->

Now in reality, transactions don't vary in size like a rock, but the sieve analogy illustrates how many accepted transactions are in the end calculated.

Below, you can see how an incorrect configuration of your SDK and server-side sampling rates would affect the transactions received. The sieves are reversed, and the SDK allows for a quarter of the total number of transactions to pass through. In this analogy, the coarser sieve now assigned for server-side sampling, wouldn't filter any more transactions because it's coarser than its counterpart. To summarize, the sampling rate configured for your SDK needs to be equal or greater than the rate set on Sentry servers.

<!-- prettier-ignore-start -->

| Sampling Process | Rate | Count |
|------------------|------|-------|
|Total transactions observed | N/A | 1000 |
| Sentry SDK sampled | 25% | 250 |
| Sentry servers sampled | 50% | 250 |

<!-- prettier-ignore-end -->

## Traces & Propagation of Sampling Decisions

TBA

## Server-Side Sampling Configurations

> <!-- TO DO: Decide if this should be a child page or remain on this page -->
> 

Author: ```suggestion
```

---

<Note>

This feature is in beta and is only available if your organization is participating in its limited release. Features in beta are still in-progress and may have bugs. We recognize the irony. If you’re interested in participating, [join the waitlist](https://sentry.io/for/performance/#updates-signup).

</Note>

Capturing a single trace involves minimal overhead, but capturing traces for every page load or every API request may add an undesirable load to your system. Sampling your events allows you to better manage the number of events sent to Sentry, so you can tailor your volume to your organization's needs. 

_Server-side sampling_ lets you control which transactions Sentry retains by setting sampling rules and rates, so you see more of the transactions you want to explore further, and less of the ones you don’t. In contrast to [client-side sampling](/platform-redirect/?next=/configuration/sampling/), server-side sampling allows you to manage your sampling configuration centrally and lets you make changes without the need to reconfigure the SDK and redeploy your application.

![The Server-side Sampling page with a list of rules](server-side-sampling.png)

> > [A trace](/product/sentry-basics/tracing/) represents the record of an entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and the trace can be distributed across multiple services or projects.

Author: ```suggestion
> [A trace](/product/sentry-basics/tracing/distributed-tracing/) represents the record of an entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and the trace can be distributed across multiple services or projects.
```

> > [A trace](/product/sentry-basics/tracing/) A trace represents the record of the entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and it can be distributed across multiple services or projects.

Author: ```suggestion
> [A trace](/product/sentry-basics/tracing/) represents the record of an entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and the trace can be distributed across multiple services or projects.
```

Because every project is different, you can set multiple sample rules with different sample rates per project. For example, you may need more events from high converting pages, critical API endpoints, or need to focus on latency issues from your latest release.

Server-side sampling also makes your SDK and your sampling configuration largely independent of each other. This helps if you're building native applications with many releases active at a given time. 

## How Sentry Samples

> Let's imagine for a moment that the total number of transactions the SDK observes is a bucket of rocks. The way Sentry samples transactions is similar to a person sifting rocks: by pouring the bucket of rocks into a sieve, you're allowing rocks small enough to pass through it:

Author: ```suggestion
Before understanding how Sentry samples, there's some concepts you need to learn:

- Total transaction volume: Total number of transactions the Sentry SDK has observed before any sampling has begun.
- Client-Side(SDK) sampling: The process of selecting which transactions are discarded or sent to Sentry for processing
- Server-Side sampling: The process of selecting which transactions are indexed after being processed.

Whether you’re sampling Client-Side or Server-Side, it’s always relative to the total transaction volume. Using the first illustration below as an example, we start the process of Client-Side sampling with a total transaction volume of 4. The Client-Side sample rate is set at 50% which leaves only 2 transactions being sent to Sentry for processing. Server-Side Sampling then takes these 2 processed transactions and indexes only 1 because it's sampling at 25% (or a quarter) of the total transaction volume.

```

> ![Visualization of how Sentry samples from total transaction volume down to post server sampling volume](how-sentry-samples.png)

Author: ```suggestion
![Visualization of how Sentry samples from total transaction volume down to post server sampling volume](how-sentry-samples.png)

<!-- to be updated -->
```

| Sampling Process | Rate | Count |
|------------------|------|-------|
|Total transactions observed | N/A | 1000 |
| Sentry SDK sampled | 50% | 500 |
| Sentry servers sampled | 25% | 250 |

<!-- prettier-ignore-end -->

Now in reality, transactions don't vary in size like a rock, but the sieve analogy illustrates how many accepted transactions are in the end calculated.

Below, you can see how an incorrect configuration of your SDK and server-side sampling rates would affect the transactions received. The sieves are reversed, and the SDK allows for a quarter of the total number of transactions to pass through. In this analogy, the coarser sieve now assigned for server-side sampling, wouldn't filter any more transactions because it's coarser than its counterpart. To summarize, the sampling rate configured for your SDK needs to be equal or greater than the rate set on Sentry servers.

<!-- prettier-ignore-start -->

| Sampling Process | Rate | Count |
|------------------|------|-------|
|Total transactions observed | N/A | 1000 |
| Sentry SDK sampled | 25% | 250 |
| Sentry servers sampled | 50% | 250 |

<!-- prettier-ignore-end -->

## Traces & Propagation of Sampling Decisions

> TBA

Reviewer 2: How about this? @jan-auer could you please sanity check this as well? Or if you have better suggestions on how to describe this part it would be very much appreciated!

For Dynamic Sampling, we want to make sampling decisions for entire traces. Sampling transactions using transaction traces is advantageous when trying to understand transactions in the context of the global system. In this case we are not interested in seeing a particular transaction, but the whole group of transactions in a trace, so sampling should occur at the trace level.

When sampling traces, the sampling rules are based on the trace context, this means that the sampling decisions are based on the information extracted from the system that initiated the transaction.

As a mental model: The head transaction in a trace determines the trace context for all following transactions in that trace. No information can be changed, added or deleted after the first propagation. The trace context is bound to only one particular trace, and all the transactions that are part of this trace.

> ## Server-Side Sampling Configurations

Reviewer 2: @imatwawana, @Jesse-Box  Reading through the whole thing now, I'm wondering if it's not worth putting this as a separate sub-section as well after the Getting Started. I think it's important enough that you might want to go straight to it and it's easier to find if it's also in the sidebar 🤔

Author: Done

![The Server-side Sampling page with a list of rules](server-side-sampling.png)

> [A trace](/product/sentry-basics/tracing/distributed-tracing/) represents the record of an entire operation you want to measure or track, like page load, an instance of a user completing some action in your application, or a cron job in your backend. A trace consists of one or more transactions and the trace can be distributed across multiple services or projects.

Because every project is different, you can set multiple sample rules with different sample rates per project. For example, you may need more events from high converting pages, critical API endpoints, or need to focus on latency issues from your latest release.

Server-side sampling also makes your SDK and your sampling configuration largely independent of each other. This helps if you're building native applications with many releases active at a given time.

## How Sentry Samples

Before understanding how Sentry samples, there are some concepts you need to know:

- Total transaction volume: Total number of transactions the Sentry SDK has observed before any sampling has begun.

> - Client-side(SDK) sampling: The process of selecting which transactions are discarded or sent to Sentry for processing.
> - Server-Side sampling: The process of selecting which transactions are indexed after being processed.

Reviewer 4: ```suggestion
- Client-side(SDK) sampling: The process of selecting which transactions are discarded or sent to Sentry for processing.
- Server-side sampling: The process of selecting which transactions are indexed after being processed.
```

Reviewer 5: If the two nouns connected by a hyphen is a Name of something then each word needs to be Capitalised.

Reviewer 4: I get it mate but not sure if people will know about... its a minor thing but looks inconsistent

Reviewer 5: Server-Side & Client-Side Sampling is also the finalised name, so let's only address any instances where it's a lower case s.

Reviewer 4: ok in this case the Client-Side shall have a capitalized letter, shouldn't it?

Reviewer 5: correct

> Whether you’re sampling client-side or server-side, it’s always relative to the total transaction volume. Using the first illustration below as an example, we start the process of client-side sampling with a total transaction volume of 4. The client-Side sample rate is set at 50% which leaves only 2 transactions being sent to Sentry for processing. Server-side sampling then takes these 2 processed transactions and indexes only 1 of them because it's sampling at 25% (or a quarter) of the total transaction volume.

Reviewer 4: ```suggestion
Whether you’re sampling client-side or server-side, it’s always relative to the total transaction volume. Using the first illustration below as an example, we start the process of client-side sampling with a total transaction volume of 4. The client-side sample rate is set at 50% which leaves only 2 transactions being sent to Sentry for processing. Server-side sampling then takes these 2 processed transactions and indexes only 1 of them because it's sampling at 25% (or a quarter) of the total transaction volume.
```

![Visualization of how Sentry samples with an effective sampling configuration](server-side-sampling-a.png)

In the next illustration, we show what would happen if the client-side sample rate was lower than the server-side rate. We again start the process of client-side sampling with a total transaction volume of 4. The client-side sample rate is lower this time, set at 25%, leaving only 1 transaction being sent to Sentry for processing. At this stage, you might be asking, “how can server-side sample 50% of the total transaction volume when the client has already discarded 75%?”. It can’t. So even though the server-side sample rate is set to 50% of the total volume, it will only index the 1 transaction it received from the client.

Ultimately, your server-side sample rate must always be equal or lower than your client-side sample rate.

![Visualization of how Sentry samples with an ineffective sampling configuration](server-side-sampling-b.png)

## Traces & Propagation of Sampling Decisions

> For Dynamic Sampling, we want to make sampling decisions for entire traces. Sampling transactions using transaction traces is advantageous when trying to understand transactions in the context of the global system. In this case we are not interested in seeing a particular transaction, but the whole group of transactions in a trace, so sampling should occur at the trace level.
> 
> When sampling traces, the sampling rules are based on the trace context, this means that the sampling decisions are based on the information extracted from the system that initiated the transaction.
> 
> As a mental model: The head transaction in a trace determines the trace context for all following transactions in that trace. No information can be changed, added or deleted after the first propagation. The trace context is bound to only one particular trace, and all the transactions that are part of this trace.
> 

Author: @jan-auer - can you review for accuracy. I've also made some suggestions for edits below: 

```suggestion
When Sentry makes sampling decisions, we do so for entire traces. Instead of looking at a particular transaction, we look at the entire group of transactions in a trace. Sampling transactions using transaction traces is ideal because it allows you to understand transactions in the context of your whole system. 

Since we sample traces, sampling rules are based on the trace context. This means that sampling decisions are based on the information extracted from the _system that initiated the transaction_. The head transaction in a trace determines the trace context for all following transactions in that trace. No information can be changed, added, or deleted after the first propagation. The trace context is bound to only one particular trace, and all the transactions that are part of this trace.

```
